### PR TITLE
DHCP4: Allow lease time to be set when missing from offer

### LINKED ISSUE #9610: systemd-networkd: [Client] Ignores DHCP Offer without lease
**systemd version the issue has been seen with**
> 238

**Used distribution**
 > CoreOS 1745.7.0

**Expected behaviour you didn't see**
 > DHCP Client acquires address from DHCP server

**Unexpected behaviour you saw**
 > DHCP Offer is ignored

screenshot_20180717_121249
screenshot_20180717_115748

This should not ignore the offer if a default lifetime is configured like this
```
[Address]
PreferredLifetime = "forever"
```

The affected code is here:
https://github.com/systemd/systemd/blob/master/src/libsystemd-network/sd-dhcp-client.c#L1291

**Steps to reproduce the problem**

Use systemd-networkd on a client system and configure the DHCP server not to send out a lease lifetime.

### PR DESCRIPTION
**Pull request by ssahani**
closes #9610

### CODE REVIEW

File: src/network/networkd-dhcp4.c [hunk 1]
```diff
@@ -1744,6 +1750,40 @@ int config_parse_dhcp_ip_service_type(
...
+        Network *network = userdata;
+        unsigned k;
+
+        assert(filename);
+        assert(section);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        /* We accept only "forever", "infinity", or "0". */
```
Comment (keszybz) [inline src/network/networkd-dhcp4.c]: This comment seems inconsistent with the code.
And there's no way to unset the setting...

**Comment by ssahani** [comment]
Updated thanks for the review.

**Comment by poettering** [comment]
hmm, i don't get this one. THis is apparently a boolean but not literally (it doesn't take yes/bo), but the option naming kinda suggests this would be a time, so why is this not in time units?

**Comment by ssahani** [comment]
> hmm, i don't get this one. THis is apparently a boolean but not literally (it doesn't take yes/bo), but the option naming kinda suggests this would be a time, so why is this not in time units?

This is a special case. In the DHCP offer there is no lease time .  It will expire after sometime and  will renew again which again gives a lease with no time (missing) . Hence there is no need to  renew since no lease time there.

File: src/network/networkd-dhcp4.c [hunk 2]
```diff
@@ -1795,6 +1801,44 @@ int config_parse_dhcp_mud_url(
...
+        assert(data);
+
+        if (isempty(rvalue)) {
+                network->dhcp_lease_lifetime = 0;
+                return 0;
+        }
+
+        /* We accept only "forever" or "infinity". */
+        if (STR_IN_SET(rvalue, "forever", "infinity"))
+                k = CACHE_INFO_INFINITY_LIFE_TIME;
```
Comment (poettering) [inline src/network/networkd-dhcp4.c:1709]: if only two values are valid here, i think we should store this as a bool, to make this clear.

File: src/network/networkd-dhcp4.c [hunk 3]
```diff
@@ -1795,6 +1801,44 @@ int config_parse_dhcp_mud_url(
...
+        else {
+                log_syntax(unit, LOG_ERR, filename, line, 0,
+                           "Invalid LeaseLifetime= value, ignoring: %s", rvalue);
+                return 0;
+        }
+
+        network->dhcp_lease_lifetime = k;
```
Comment (poettering) [inline src/network/networkd-dhcp4.c]: there also needs to be a way to turn this off again.

**Comment by ssahani** [comment]
> i still don' get this II must say. So i understood that this is about receiving a lease offer from a DHCP server lacking a lease timeout. So what precisely did we previously do in this case, and what will we do now? why is this configurable precisely, i.e. if i use hthis and if i don't, what are my reasons for these two cases?

The lease time missing from the lease as reported in the issue #9610 . Right now we don#t configure it and sends again a  discoverer  . Now the lease is permanently be there like no expiry more like static. 

> 
> if the lease is infinity, when is the lease refreshed? never? Only by FORCERENEW? Can you elaborate?

 That's right. It 's never refreshed since server does not offer a lease time at all and by `FORCERENEW`  only

**Comment by poettering** [comment]
But shouldn't it be a boolean option then, maybe called `AcceptMissingLeaseTimeAsForever=" or so?

Is there any RFC saying what to do in this case?

**Comment by ssahani** [comment]
> But shouldn't it be a boolean option then, maybe called `AcceptMissingLeaseTimeAsForever=" or so?

yes make sense.
> 
> Is there any RFC saying what to do in this case?

I cant find anything from RFC.  But what closer is this piece of information.
```
The client may ask for a
   permanent assignment by asking for an infinite lease.  Even when
   assigning "permanent" addresses, a server may choose to give out
   lengthy but non-infinite leases to allow detection of the fact that
   the client has been retired.
```

**Comment by keszybz** [comment]
> if only two values are valid here, i think we should store this as a bool, to make this clear.

Actually, I think current interface is OK. It only accepts LeaseLifetime=forever, but it leaves the door open to some other values in the future if we want it. So I think part is fine as is.

But this part still needs fixign (https://github.com/systemd/systemd/pull/15274/files#r425681491):
> there also needs to be a way to turn this off again.

Just make LeaseLifetime= unset the setting.

**Comment by poettering** [comment]
hmm, but if accepts other (time) values it should probably be called LeaseLifetimeSec=, i.e. to indicate the unit when numeric values are specified, no?

Also, the question is what it would do: would it override whatever lifetime is specified in the lease, or would it only be the fallback if no lifetime is specified? I think the option the way it is currently implemented is only about the latter, but I am very sure the option should somehow communicate this in its name. i.e. FallbackLeaseLifetimeSec= or DefaultLeaseLifetimeSec= or so.

**Comment by keszybz** [comment]
True.  FallbackLeaseLifetimeSec= or DefaultLeaseLifetimeSec=  would be better.

**Comment by ssahani** [comment]
I have updated with `FallbackLeaseLifetimeSec=` now. Thanks for the suggestions.

**Pull request merged**